File: Python/perfect_number.py
# https://leetcode.com/problems/perfect-number/
from utils.examination import Examiner, Testcase, Logger


class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        if 1 == num:
            return False
        i: int = 2
        max_: int = num
        sum_: int = 1
        while max_ > i:
            if 0 == num % i:
                sum_ += i + num // i
                if sum_ > num:
                    return False
                max_ = num // i
            i += 1
        return sum_ == num
def examine()->None:
    Examiner(Solution(), 'checkPerfectNumber').run((
        Testcase((28,), True),
        Testcase((1,), False),
        Testcase((2,), False),
        Testcase((7,), False),
    ))

File: TypeScript/solutions/perfect-number.ts
// https://leetcode.com/problems/perfect-number/
import { Examiner, Testcase } from '../utils';


function checkPerfectNumber(num: number): boolean {
    if (num === 1) return false;
    let i = 2, max = num, sum = 1;
    while (max > i) {
        if (num % i === 0) {
            sum += i + Math.floor(num / i);
            if (sum > num) return false;
            max = Math.floor(num / i);
        }
        i++;
    }
    return sum === num;
}

export function examine(): void {
    new Examiner(checkPerfectNumber).run([
        new Testcase([28], true),
        new Testcase([1], false),
        new Testcase([2], false),
        new Testcase([7], false),
    ]);
}
